demos/multi_material/compositional_buoyancy/compositional_buoyancy.py

```diff
@@ -158,101 +156,106 @@
 approximation = BoussinesqApproximation(Ra, RaB=RaB)
 # -
 
+# Let us now verify that the material fields have been correctly initialised. We plot
+# the compositional Rayleigh number across the domain.
+
+# + tags=["active-ipynb"]
+# fig, axes = plt.subplots()
+# axes.set_aspect("equal")
+# contours = tricontourf(
+#     Function(psi).interpolate(RaB), levels=linspace(0.0, 1.0, 11), axes=axes
+# )
+# fig.colorbar(contours, label="Compositional Rayleigh number")
+# -
+
 # As with the previous examples, we set up an instance of the `TimestepAdaptor` class
 # for controlling the time-step length (via a CFL criterion) whilst the simulation
 # advances in time. We specify the initial time, initial time step $\Delta t$, and
 # output frequency (in time units).
 
-time_now = 0  # Initial time
-delta_t = Function(R).assign(1)  # Initial time step
-output_frequency = 10  # Frequency (based on simulation time) at which to output
+time_now = 0.0  # Initial time
+time_step = Function(R).assign(1.0)  # Initial time step
+output_frequency = 10.0  # Frequency (based on simulation time) at which to output
 t_adapt = TimestepAdaptor(
-    delta_t, u, V, target_cfl=0.6, maximum_timestep=output_frequency
+    time_step, u, V, target_cfl=0.6, maximum_timestep=output_frequency
 )  # Current level-set advection requires a CFL condition that should not exceed 0.6.
 
-# This problem setup has a constant pressure nullspace, which corresponds to the
-# default case handled in G-ADOPT.
-
-Z_nullspace = create_stokes_nullspace(Z)
+# Here, we set up the variational problem for the Stokes and level-set systems. The
+# former depends on the approximation defined above, and the latter includes both
+# advection and reinitialisation components.
 
-# Boundary conditions are specified next: no slip at the top and bottom and free slip
-# on the left and ride sides. No boundary conditions are required for level set, as the
-# numerical domain is closed.
+# +
+# This problem setup has a constant pressure nullspace, which corresponds to the default
+# case handled in G-ADOPT.
+stokes_nullspace = create_stokes_nullspace(Z)
 
+# Boundary conditions for the Stokes system: no slip at the top and bottom and free slip
+# on the left and right sides.
 stokes_bcs = {
-    boundary.bottom: {"u": 0},
-    boundary.top: {"u": 0},
-    boundary.left: {"ux": 0},
-    boundary.right: {"ux": 0},
+    boundary.bottom: {"u": 0.0},
+    boundary.top: {"u": 0.0},
+    boundary.left: {"ux": 0.0},
+    boundary.right: {"ux": 0.0},
 }
+# Instantiate a solver object for the Stokes system and perform a solve to obtain
+# initial pressure and velocity.
+stokes_solver = StokesSolver(
+    stokes,
+    T,
+    approximation,
+    bcs=stokes_bcs,
+    nullspace=stokes_nullspace,
+    transpose_nullspace=stokes_nullspace,
+)
+stokes_solver.solve()
```

Comment:
Why is this needed outside of the time loop? Perhaps a comment to help a user. It's different to our other cases where the solves are all inside the time loop.

Reply:
If we specify initial conditions for all variables, then I do not see a reason not to have an initial condition for velocity and pressure as well. That is what this solver call is achieving. That also implies that all our demos produce Stokes and energy solutions that are out of sync, right?

Reply:
I think it depends on the order of things in the other demos. In theory if the first solve inside the time loop is a stokes solve (as it is with case case), it is already getting an initial guess for u and p with the prescribed RHS, so I don't see a problem there. Can we not do the same here, rather than have this separate call?